incubator/hnc/internal/reconcilers/anchor.go

```diff
@@ -161,69 +163,70 @@ func (r *AnchorReconciler) shouldDeleteSubns(log logr.Logger, inst *api.Subnames
-	// doesn't have any children itself, or if cascading deletion is enabled.
+	// The namespace is in a good state and is linked to this anchor, so it can be deleted. For
+	// safety's sake, we only actually delete subnamespaces if they're leaves or if ACD=true on one of
+	// their ancestors; the webhooks *should* ensure that this is always true before we get here, but
+	// if something slips by, we'll just leave the old subnamespace alone with a missing-anchor
+	// condition.
+	cns := r.forest.Get(inst.Name)
 	return cns.ChildNames() == nil || cns.AllowsCascadingDelete()
 }
 
-func (r *AnchorReconciler) removeFinalizers(log logr.Logger, inst *api.SubnamespaceAnchor, snsInst *corev1.Namespace) bool {
-	pnm := inst.Namespace
-	sOf := snsInst.GetAnnotations()[api.SubnamespaceOf]
+// shouldFinalizerAnchor determines whether the anchor is safe to delete. It should only be called once
+// we know that we don't need to delete the subnamespace itself (e.g. it's already gone, it can't be
+// deleted, it's in the process of being deleted, etc).
+func (r *AnchorReconciler) shouldFinalizerAnchor(log logr.Logger, inst *api.SubnamespaceAnchor, snsInst *corev1.Namespace) bool {
+	// If the namespace is gone, we can finish deleting the anchor. If the namespace is present but
+	// isn't already being deleted, we assume that shouldDeleteSubns has explicitly decided *not* to
+	// delete it, which means that there's nothing stopping the anchor from being deleted.
+	if snsInst.Name == "" || snsInst.DeletionTimestamp.IsZero() {
+		return true
+	}
 
-	// This switch statement has an explicit case for all conditions when we *can* delete the
-	// finalizers. The default behaviour is that we *cannot* delete the finalizers yet.
-	switch {
-	case snsInst.Name == "":
-		log.Info("The subnamespace is already purged; allowing the anchor to be deleted")
-	case sOf != pnm:
-		log.Info("The subnamespace believes it is the subnamespace of another namespace; allowing the anchor to be deleted", "annotation", sOf)
-	case snsInst.DeletionTimestamp.IsZero():
-		// onDeleting has decided not to try to delete this namespace. Either cascading deletion isn't
-		// allowed (and the user has bypassed the webhook), or the CRD has been deleted and we don't
-		// want to invoke cascading deletion. Just allow the anchor to be deleted.
-		log.Info("We've decided not to delete the subnamespace; allowing the anchor to be deleted")
-	default:
+	switch inst.Status.State {
+	case api.Ok:
+		// The anchor is in a good state, and the namespace exists and is being deleted. Keep waiting.
 		return false
+	case api.Missing:
+		// This should never happen; we know that the namespace exists since it's name is non-empty, and
+		// we *just* updated the state. Log an error but allow the object to be deleted anyway - a
+		// missing anchor isn't the end of the world.
+		log.Error(errors.New("Illegal state"), "Anchor in missing state but namespace exists (should be ok or conflict")
```

Comment:
I think for `missing` it's  means the subns is deleted and we should remove anchor finalizers (agreed on `return true`) and log no error.

Reply:
But if the subns were already deleted, then on line 206 above, `snsInst.Name` would have been empty and we wouldn't get here. That's why I don't think it's possible to get to this line and still have the state be `Missing`. Does that sound right?